core-tests/src/test/java/org/mule/runtime/core/internal/util/queue/LocalTxQueueTransactionTestCase.java

```diff
@@ -155,10 +202,95 @@ public void offerAndFailBetweenRealOfferAndCommitThenRecover() throws Exception
     }
     txLog.close();
     txLog = new TestTransactionLogger(temporaryFolder.getRoot().getAbsolutePath(), muleContext);
-    queueTransactionRecoverer = new LocalTxQueueTransactionRecoverer(txLog, createQueueProvider(outQueue));
+    queueTransactionRecoverer = new LocalTxQueueTransactionRecoverer(txLog, createRecoverOnlyQueueProvider(outQueue));
     queueTransactionRecoverer.recover();
     Serializable muleEvent = outQueue.poll(TIMEOUT);
     assertThat(muleEvent, nullValue());
   }
 
+  @Test
+  public void pollDoesNotReturnNullBeforeTimeoutUsingTransactionContext() throws InterruptedException {
+    // When input queue is empty
+    assertThat(inQueue.getSize(), is(0));
+
+    long timeout = 300;
+
+    // If we poll an element from the queue with a given timeout
+    long timeBeforePoll = currentTimeMillis();
+    Serializable polledValue = persistentTransactionContext.poll(inQueue, timeout);
+    long timeAfterPoll = currentTimeMillis();
+
+    // Then the value is null
+    assertThat(polledValue, nullValue());
+    // And it waited at least for the given timeout
+    assertThat(timeAfterPoll - timeBeforePoll, greaterThanOrEqualTo(timeout));
```

Comment:
use some kind of error margin for this, `currentTimeMillis` has some precision issues that may make this assertion flaky. Check out `org.hamcrest.number.IsCloseTo.closeTo `